### PR TITLE
ENH: Optimize nrows in read_excel

### LINKED ISSUE #32727: read_excel opimize nrows
#### Code Sample

```python
pd.read_excel(fname, nrows=10)
```
#### Problem description

Pandas has option toread only several rows of excel files.
But now it always read all rows and after pandas cut some part. 
For example, file have 100 columns and 50k rows, but for test need only first 10 rows.
Now pandas will read to list all 50k rows which use memory and take too many time to read.

les this should explain **why** the current behaviour is a problem and why the expected output is a better solution.]

#### Expected Output

Better solution should be read only rows which need for operation.

as I understand there shoul be some changes

pandas/io/excel/_base.py
```python
    @abc.abstractmethod
    def get_sheet_data(self, sheet, convert_float, header, skiprows, nrows):
        pass
```
pandas/io/excel/_base.py
```python
   data = self.get_sheet_data(sheet, convert_float, header, skiprows, nrows)
```

and in files _openpyxl.py, _odfreader.py, _xlrd.py
there should be something like
```python

    def get_sheet_data(self, sheet, convert_float: bool, header: int, skiprows: int, nrows: int) -> List[List[Scalar]]:
        data = []  # type: List[List[Scalar]]
        skiprows = 0 if skiprows is None else skiprows
        header = 0 if header is None else header

        for row in sheet.rows:
            if nrows is not None:
                if header > 0:
                    #print("skip lines before header")
                    header -= 1
                    data.append(["", ""])
                    continue
    
                if skiprows > 0:
                    #print("skip skiprows after header")
                    skiprows -= 1
                    data.append(["", ""])
                    continue
    
                if nrows >= 0:
                    #print("read nrows")
                    nrows -= 1
                else:
                    break
            data.append([self._convert_cell(cell, convert_float) for cell in row])

        return data
```

With this changes read_excel with engine='openpyxl' takes only 5 seconds instead of 50 seconds of current version. And if file will contain 1kk rows, it will take always around 5 seconds, but current version will take tens of minutes.

### PR DESCRIPTION
**Pull request by mproszewska**
- [x] closes #32727
- [x] tests passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] If header, skiprows and nrows are integers, rows that will be skipped are not loaded in get_sheet_data function.

### CODE REVIEW

**Comment by mproszewska** [comment]
Tests which are not connected to that PR fail.

File: pandas/io/excel/_base.py [hunk 1]
```diff
@@ -380,7 +380,7 @@ def get_sheet_by_index(self, index):
         pass
 
     @abc.abstractmethod
-    def get_sheet_data(self, sheet, convert_float):
+    def get_sheet_data(self, sheet, convert_float, header, skiprows, nrows):
```
Comment (WillAyd) [inline pandas/io/excel/_base.py:383]:
```suggestion
    def get_sheet_data(self, sheet, convert_float, header: int, skiprows: int, nrows: int):
```

We try to annotate all new arguments, so would be good to do here (I'm assuming they are all int)

**Review by WillAyd — CHANGES_REQUESTED** [review]
Thanks for the PR. Can you add tests?

File: pandas/io/excel/_base.py (hunk 1 above)
Comment (mproszewska) [reply pandas/io/excel/_base.py:383]: Both header and skiprows can be a list, but this optimization works only if all three variables are ints. nrows is always int. Should I annotate only nrows?

Comment (WillAyd) [reply pandas/io/excel/_base.py:383]: If you can annotate everything new that you’ve added with the appropriate types that would be helpful. Sounds like the others should be `Union[int, Sequence[int]]` then

Comment (mproszewska) [reply pandas/io/excel/_base.py:383]: Done

**Comment by mproszewska** [comment]
But tests for readers already exist and include header, skiprows and nrows variables. Should I add something more?

File: pandas/io/excel/_xlrd.py [hunk 2]
```diff
@@ -95,9 +105,30 @@ def _parse_cell(cell_contents, cell_typ):
                     cell_contents = val
             return cell_contents
 
-        data = []
+        data: List[List[Scalar]] = []
 
+        if nrows is not None:
```
Comment (mroeschke) [inline pandas/io/excel/_xlrd.py:110]: This block of code looks repeated in the various excel files. Could we turn it into a shared function?

File: pandas/io/excel/_xlrd.py [hunk 3]
```diff
@@ -95,9 +105,30 @@ def _parse_cell(cell_contents, cell_typ):
...
+            _validate_integer("nrows", nrows)
+        header = 0 if header is None else header
+        skiprows = 0 if skiprows is None else skiprows
+        if isinstance(header, list) or isinstance(skiprows, list):
+            nrows = None
         for i in range(sheet.nrows):
+
+            if nrows is not None:
```
Comment (mroeschke) [inline pandas/io/excel/_xlrd.py:118]: This block of code looks repeated in the various excel files. Could we turn it into a shared function?

Comment (mproszewska) [reply pandas/io/excel/_xlrd.py:117]: Opened excel file is won't be closed if exception associated with the value of nrows is raised in `get_sheet_data` methods, so I moved `_validate_integer` to `read_excel` method in `_base.py` file.

**Comment by alimcmaster1** [comment]
Mind fixing up the code checks @mproszewska and I agree with @mroeschke comments we should find a way to get rid of duplicate code

File: pandas/io/excel/_xlrd.py (hunk 3 above)
Comment (mroeschke) [reply pandas/io/excel/_xlrd.py:117]: Right. But we could create another function in `_base.py` that generates flags to `continue` or `break` bases on this repeated code. 

Same for the block you added above

```
        header = 0 if header is None else header
        skiprows = 0 if skiprows is None else skiprows
        if isinstance(header, list) or isinstance(skiprows, list):
            nrows = None
```

File: pandas/io/excel/_base.py [hunk 4]
```diff
@@ -383,9 +386,18 @@ def get_sheet_by_index(self, index):
...
         pass
 
+    def should_read_row(self, index, header, skiprows, nrows):
```
Comment (mroeschke) [inline pandas/io/excel/_base.py:392]: Could you give this function a docstring?

**Comment by mproszewska** [comment]
I added suggested changes, but one test is still not passing and I'm not sure how to fix it.

File: pandas/io/excel/_base.py [hunk 5]
```diff
@@ -383,9 +383,37 @@ def get_sheet_by_index(self, index):
...
```
Comment (jreback) [inline pandas/io/excel/_base.py:389]: can you type the args & return type

**Review by jreback — CHANGES_REQUESTED** [review]
- can you add some tests for each backend which are exercised by this.
- update types
- add a whatsnew in perf section.

File: doc/source/whatsnew/v1.1.0.rst [hunk 6]
```diff
@@ -427,6 +427,7 @@ Performance improvements
   :meth:`DataFrame.sparse.from_spmatrix` constructor (:issue:`32821`,
   :issue:`32825`,  :issue:`32826`, :issue:`32856`, :issue:`32858`).
 - Performance improvement in reductions (sum, prod, min, max) for nullable (integer and boolean) dtypes (:issue:`30982`, :issue:`33261`, :issue:`33442`).
+- Performance improvement in `read_excel` for integer header, skiprows and nrows (:issue:`33281`).
```
Comment (gfyoung) [inline doc/source/whatsnew/v1.1.0.rst:430]:
```suggestion
- Performance improvement in `read_excel` for integer ``header``, ``skiprows``, and ``nrows`` (:issue:`33281`).
```

File: pandas/io/excel/_base.py [hunk 7]
```diff
@@ -383,9 +385,43 @@ def get_sheet_by_index(self, index):
         pass
 
     @abc.abstractmethod
-    def get_sheet_data(self, sheet, convert_float):
+    def get_sheet_data(self, sheet, convert_float, header, skiprows, nrows):
         pass
 
+    def should_read_row(
+        self,
+        index,
```
Comment (gfyoung) [inline pandas/io/excel/_base.py:393]: Type this parameter as well?

File: pandas/tests/io/excel/test_readers.py [hunk 8]
```diff
@@ -1084,3 +1084,17 @@ def test_excel_high_surrogate(self, engine):
...
+            DataFrame(data.iloc[3:6])
+            .reset_index(drop=True)
+            .rename(columns=data.iloc[2].rename(None))
+        )
+        actual = pd.read_excel(
+            "test1" + read_ext, engine=engine, header=1, skiprows=2, nrows=3
+        )
+        tm.assert_frame_equal(expected, actual)
+        actual = pd.read_excel("test1" + read_ext, engine=engine, skiprows=3, nrows=3)
+        tm.assert_frame_equal(expected, actual)
```
Comment (gfyoung) [inline pandas/tests/io/excel/test_readers.py:1100]: Parameterize this test over the `kwargs` passed into `read_excel`.

**Comment by mproszewska** [comment]
I removed skipping rows in openpyxl, because it seems like there is a problem with files leaking. It might be connected to [this](https://bitbucket.org/openpyxl/openpyxl/issues/673/workbook-context-manager#comment-36565768).

File: pandas/core/sorting.py [hunk 9]
```diff
@@ -385,7 +385,7 @@ def ensure_key_mapped(values, key: Optional[Callable], levels=None):
     from pandas.core.indexes.api import Index
 
     if not key:
-        return values.copy()
+        return values
```
Comment (jreback) [inline pandas/core/sorting.py:388]: why is this being changed ?

File: pandas/io/excel/_base.py [hunk 10]
```diff
@@ -439,7 +482,16 @@ def parse(
             else:  # assume an integer if not a string
                 sheet = self.get_sheet_by_index(asheetname)
 
-            data = self.get_sheet_data(sheet, convert_float)
+            gsd_header = 0 if header is None else header
```
Comment (jreback) [inline pandas/io/excel/_base.py:485]: what does gsd mean? please make meaningful variable names

File: pandas/io/excel/_odfreader.py [hunk 11]
```diff
@@ -80,6 +87,16 @@ def get_sheet_data(self, sheet, convert_float: bool) -> List[List[Scalar]]:
         table: List[List[Scalar]] = []
 
         for i, sheet_row in enumerate(sheet_rows):
+
+            should_continue, should_break = self.should_skip_row(
```
Comment (jreback) [inline pandas/io/excel/_odfreader.py:91]: I am not sure this interface is obvious at all. prefer simple routines

if should_skip_rows(...):
....

when is a break condition? can you just do it here?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/excel/_base.py (hunk 10 above)
Comment (mproszewska) [reply pandas/io/excel/_base.py:485]: it's supposed to be short for get_sheet_data, but I guess it's not clear

File: pandas/io/excel/_odfreader.py (hunk 11 above)
Comment (mproszewska) [reply pandas/io/excel/_odfreader.py]: I changed it. Break condition is now handled differently

File: pandas/core/sorting.py (hunk 9 above)
Comment (mproszewska) [reply pandas/core/sorting.py]: problem with merging branches. sorry for that

**Comment by WillAyd** [comment]
Can you add a benchmark for this to show performance improvement?

**Comment by WillAyd** [comment]
@mproszewska can you post the ASV results here as a comment?

**Comment by mproszewska** [comment]
```
[ 75.00%] ··· io.excel.ReadExcel.time_read_excel_nrows                                   ok
[ 75.00%] ··· ========== ===========
                engine              
              ---------- -----------
                 xlrd     1.92±0.1s 
               openpyxl   4.69±0.5s 
                 odf       11.3±1s  
              ========== ===========

[ 75.00%] · For pandas commit f0a4e8e5 <master> (round 2/2):
[ 75.00%] ·· Building for conda-py3.6-Cython0.29.16-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt...
[ 75.00%] ·· Benchmarking conda-py3.6-Cython0.29.16-matplotlib-numba-numexpr-numpy-odfpy-openpyxl-pytables-pytest-scipy-sqlalchemy-xlrd-xlsxwriter-xlwt
[100.00%] ··· Setting up io/excel.py:62                                                  ok
[100.00%] ··· io.excel.ReadExcel.time_read_excel_nrows                                   ok
[100.00%] ··· ========== ============
                engine               
              ---------- ------------
                 xlrd     2.11±0.08s 
               openpyxl   4.65±0.5s  
                 odf      14.5±0.1s  
              ========== ============
```

**Comment by WillAyd** [comment]
Can you fix merge master and fix merge conflicts? That may also fix CI

Can you post the output of a continuous asv run instead of just the snippet? It should provide results across multiple runs at the end

**Comment by simonjayhawkins** [comment]
@mproszewska can you address comments?

**Comment by WillAyd** [comment]
Closing as I think this is stale but @mproszewska ping if you'd like to pick back up and can fix merge conflicts

**Pull request closed without merging**